Hibernate 5.4 extension updates

michaelborn opened the pull request:
# Hibernate 5.4 Extension Updates

Thanks to a client project, we (Ortus Solutions) have had the opportunity to assist with closing some of the open ORM / Hibernate extension issues, as well as bring the extension up to modern standards in a few areas. Notably, this PR:

1. resolves [a number of bugs in the current Hibernate extension](#bug-fixes), 
2. adds [new features](#new-features), 
3. cuts the ORM reload time _in half_ via [refactoring the initialization process](#initialization-rework), 
4. adds [documentation via the readme and javadocs](#documentation), 
5. [adds or improves the existing test suite](#tests), 
6. and does some general code cleanup and [housekeeping](#housekeeping)

Explanation of changes:

## Bug Fixes

* Fix [LDEV-4017](https://luceeserver.atlassian.net/browse/LDEV-4017) (cbf99de7076ce0c5aa6e6ecbe98bd1303a3c5f98)
* Fix [LDEV-4150](https://luceeserver.atlassian.net/browse/LDEV-4150) (866c11ef54390c81f801ab5bd273ed7cb571d454)
* Fix [LDEV-2049](https://luceeserver.atlassian.net/browse/LDEV-4308) (a7ab519f3deb519e6449523a82cd6b2f305d5159)

## New Features

* Implemented [LDEV-3525](https://luceeserver.atlassian.net/browse/LDEV-3525) (2fc53efbf404461cf5edd901ce537fdc017c08fd, 5fc76c0664145db8fecd0ad54582febc22697553, 681b068ba1aa60a8a8760c72dd3b030bc00431b8, 60a4f4cdcf196ae62b2192d58b7cba33827fa712, 7c61672440c13c8f55ee94e3e07215d91961a055, 57a4309e534c6dbfe03185140ba104385c50a475, a0759341512ca70d8e40072435ad8497349a01e6)

## Initialization Rework

See commits 41f0d860bc41efd0a59f381556c4d011cf14a7d1, e8d0dd0907f059aa1b669c9e1792bc0d8a97f9c1, and e31782f3d24d363397ee35154c09f1f58f053e8c

* Don't reinit ORM except 1. on first run or 2. if ORMReload() is fired
  * This fixes a bug where `INIT_CFCS` and `INIT_ALL` were mapped to the same int, so each page load would re-inspect all ORM entities.
  * Avoids re-hashing the ORM config on every request
  * Reduces unnecessary config hashing and component scans on every request
* Use `hasTempCFCs()` boolean method for clarity when looking for ORM entities in the list of pre-inspected entities
* Don't attempt to reinit when instantiating an entity

This refactor alone reduces the time to reinit 1,000 entities from ~14s to ~4s - a ~70% improvement!

See performance data collected via [Hibernate Playground](https://github.com/michaelborn/Hibernate-Playground) against the 3.5 extension, 5.4 extension, and 5.4 with modifications:

![Pivot table showing execution time for various actions on three different extension versions](https://user-images.githubusercontent.com/8106227/218541470-f9697030-a32f-4874-8af3-bd29d758d368.png)

## Documentation

* New project README for (cd7b05d8ad981ee201834982682f39b7e121dfd0)
* Added automatic javadocs generation via Maven and GitHub Actions (22350751fbaacb13f860c643cc9e046b55ef9ce2)
* Added javadocs / documentation blocks ( 888a2cb7ec240f3d20e3a5641350b0e50cf02765, 8174c014a72649bb10455c4bee92faf85f1ed439, a5ad71bf794557b4974b13325a2832c7e60b3f2d, d553f6e7a7cde86f7f111adfd22464586f270cfd, 032e07d8ecae43e67abb81e95b6cb4997a8d9219, 2198aa10c289025ad4c03cb2c55b1bab1da61056, 69854bb0f85229d173709b8674b86445810e6011, d9887461a5424fd23b7181aec45106a44830b870, 946b1bf5dee07a6707ef96ae4bfd088be89a1125)

## Tests

* Imported tests for LDEV-4150 (20a0005708d64e18596f54c5a3836fc0c9dd41da)
* Imported tests for LDEV-3907 (4271c5ffef05c011007efa444f4832d8fad3feb2)
* Added test for `this.ormSettings.ormConfig` XML file (c0fdc924e2e8c8223c93fb85bc67c0252b2cd3fe)
* Added test for `this.ormSettings.cacheconfig` XML file (2e1455eef10bf51f50f7bb2556c7081aa849b594)
* Added test for `datasource` annotation on individual entity components (6cac4576e19067b049b3bbff7d1e1623bc45e8da)
* Setup Junit tests (d3266be8901f6276d081d41383622faf5ce1a162, 73a99392b571763b0d53a26165f9ddf0925e6131, 70482ce6455f310848244d24f6bdd331cb16f4fb, a5f4b89346a9b440a89a6e08f904831a071d2dac, b525356def1984dd5e0cfccfeade25c8aaaa85d2)
* Added `test.sh` for easily running Lucee test cases (c997777c49880fe8b151294bac64e4c41bb62d71)

## Housekeeping

* Moved all `ORM*()`, `entity*()` methods such as `ORMReload()`, `entitySave()` to extension (e11604903771c0954340a0ed16f3e161b4c1abe3, c0879cd3c300a097ff4113a7527ea5b205d20353, 00681de42ac3f339018848c7e5780b05c2dd6828)
* Drop support for Lucee dialect / `.lucee` components (aa4aa491962706987cfa744cb99b79bb1aabd556), (ac5abbf4cf9428956fc058d04bc9f3e041c5a920)
* Drop unused methods and classes cluttering the source code (c00e275dd4c0552868a13b596f54ec6f09a09315, 96eb530a78e45e6839dd6be381601cfcd7e8166a, 461e423f445f32922f5360c48ad4d8a6d611b10f, )
* Moved all `*Util.java` classes to `.util` package (24c41154a042652eeda5dfc490e3383f3d54185b)
* Use Hibernate setting key names if they exist (4c51369f3192f4ccbdca00556c69dec304a12e49, d342d219548f47d71947b73dc61677443371789e, 0edede733ed8c03162ea882a0f248ec77181c22f)
* Refactored configuration to use a Builder pattern for clarity (c7d74a49e85722c14199e51caec303d6b8aed71b, 929fff462e3fd2ed7428a00191c45169feeaa689)
* Added automatic java source formatting via `formatter-maven-plugin` (c43c60e98195c1cb98aa2af3d81a40a1003f1df4, c4c3ad966cd85ec42110fa1643559be889274de9, 65f44cf73f10212f1609b3f8acfde43e068edb6f)

michaelborn commented:
It's worth noting that through all these changes and refactorizations, all tests pass locally on H2, MySQL, MSSQL, and Postgres, as well as on H2 in the test suite.

Latest build: https://github.com/Ortus-Solutions/extension-hibernate/actions/runs/4166673150

zspitzer commented:
@michaelborn wow, nice. is the token required for checkout for the build step (current failing), only the final format step requires the token?

michaelborn commented:
I added the token to the build to fix a permissions issue on our repo fork when uploading the test results. Since the original build worked without a token, we should be fine to remove it.

-------- Original Message --------
On Feb 13, 2023, 6:56 PM, Zac Spitzer wrote:

> ***@***.***(https://github.com/michaelborn) wow, nice. is the token required for checkout for the build step (current failing), only the final format step requires the token?
>
> —
> Reply to this email directly, [view it on GitHub](https://github.com/lucee/extension-hibernate/pull/36#issuecomment-1428879770), or [unsubscribe](https://github.com/notifications/unsubscribe-auth/AB53B42N3C2M73LIUKANE5LWXLC3FANCNFSM6AAAAAAU2TE6CI).
> You are receiving this because you were mentioned.Message ID: ***@***.***>

michaelborn commented:
@zspitzer the main build passes now. :tada:

lmajano commented:
Any comments @michaeloffner @micstriit ??

zspitzer commented:
@michaelborn can you rebase/squash some of the commits, this PR doesn't need to be 157 commits

michaelborn commented:
> @michaelborn can you rebase/squash some of the commits, this PR doesn't need to be 157 commits

Sure, @zspitzer you can squash all commits into one when you merge the PR. See screenshot:

![Screenshot from 2023-02-22 14-14-13](https://user-images.githubusercontent.com/8106227/220646926-87ddc852-6adb-4ef9-b707-6bad2b566574.png)

zspitzer commented:
@michaelborn yeah, I know about that, but i'd really like to avoid a (single huge commit / 157 commits), because it make it much harder for anyone reviewing changes in the future, this PR is already way to hard to navigate

michaelborn commented:
So, @zspitzer I've tried and failed a few times now at squashing groups of commits into individual feature/fix commits. I may be missing some esoteric GitKraken command, but right now I don't see how to pare this PR down further.

Choices as I see it:

A. Merge as is.
B. Merge squashed. (I understand your not liking this option!)
C. Close this PR and open a new PR for each feature or bug fix.

C is certainly the best option here, but I am concerned about 1) merge or rebase conflicts, and 2) sequencing issues (many features or fixes build on each other).

jclausen commented:
@zspitzer What is the concern with the number of commits in the PR?  I, personally, appreciate many commits with specific messages because GitLens and other tools for the IDE allow me to see why a particular line was changed at a particular step. Are you just looking to have commit-level visibility to the individual changes?

@michaelborn If you need to you could probably try [using `fixup` to auto-squash commits on the same file](https://fle.github.io/git-tip-keep-your-branch-clean-with-fixup-and-autosquash.html) - or consolidate all of the documentation updates.

lmajano commented:
Are we ever going to get some feedback on the PR apart that it has too many commits?  We spent a lot of time and effort on this PR.  I don't understand the issue of the many commits when you can squash and merge all the changes into the branch. I always do this on all of our repositories to maintain a single PR -> to a single commit insertion point.

The fact that we have not had feedback from the main developer really concerns us.  Can we at least get an ETA or acknowledgment that we are in the right direction?

jclausen commented:
[bump]  🦗

Nitinrajyadav commented:
[bump]

michaeloffner commented on source/java/.project:
```diff
@@ -14,4 +14,15 @@
 	<natures>
 		<nature>org.eclipse.jdt.core.javanature</nature>
 	</natures>
+	<filteredResources>
```
this is not compatible with other project settings from other projects

michaeloffner commented on source/java/src/META-INF/EXTENSION.MF:
```diff
@@ -0,0 +1 @@
+Manifest-Version: 1.0
```
What is this good for?

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/EntityDelete.java:
```diff
@@ -0,0 +1,33 @@
+/**
```
This function has multiple problems
1. this functions is already defined in the Lucee core, the Lucee core is aware of what an ORM Engine is and because also has this function, moving that function is a possibility, but then i'm not positive that you load the session in a correct way. what is the benefit to move that function outside the Lucee core, do you wanna remove that function from the Lucee core?
2. i nowhere see this function registered via fld, so i guess it is not used at all
3. a function outside the Lucee core should extend the abstract class BIF (best see the S3 extension as an example), otherwise the Lucee core has to create a wrapper on demand for that class what is time consuming.

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/EntityLoad.java:
```diff
@@ -0,0 +1,85 @@
+/**
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/EntityLoadByExample.java:
```diff
@@ -0,0 +1,38 @@
+/**
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/EntityLoadByPK.java:
```diff
@@ -0,0 +1,21 @@
+package org.lucee.extension.orm.functions;
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/EntityMerge.java:
```diff
@@ -0,0 +1,33 @@
+/**
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/EntityNameArray.java:
```diff
@@ -0,0 +1,33 @@
+/**
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/EntityNameList.java:
```diff
@@ -0,0 +1,37 @@
+/**
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/EntityNew.java:
```diff
@@ -0,0 +1,76 @@
+/**
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/EntityReload.java:
```diff
@@ -0,0 +1,34 @@
+/**
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/EntitySave.java:
```diff
@@ -0,0 +1,38 @@
+/**
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/EntityToQuery.java:
```diff
@@ -0,0 +1,39 @@
+/**
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/ORMClearSession.java:
```diff
@@ -0,0 +1,40 @@
+/**
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/ORMCloseAllSessions.java:
```diff
@@ -4,23 +4,28 @@
- * You should have received a copy of the GNU Lesser General Public 
+ *
+ * You should have received a copy of the GNU Lesser General Public
  * License along with this library.  If not, see <http://www.gnu.org/licenses/>.
```
not sure what is going on here, but print and aprint are classes present in the lucee core and every extension, they are used to make printout while developing and should NOT be removed. 
- print can be used like the dump function and can better show complex structures than System.out
- you can simply remove the "extends aprint" from the print class to see where the class is used, helpful to clean up

please do not simply remove stuff you don't completely understand the need for

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/ORMCloseSession.java:
```diff
@@ -0,0 +1,40 @@
+/**
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/ORMEvictCollection.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ *
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/ORMEvictEntity.java:
```diff
@@ -0,0 +1,41 @@
+/**
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/ORMEvictQueries.java:
```diff
@@ -0,0 +1,39 @@
+/**
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/ORMExecuteQuery.java:
```diff
@@ -0,0 +1,108 @@
+/**
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/ORMFlush.java:
```diff
@@ -0,0 +1,39 @@
+/**
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/ORMGetSession.java:
```diff
@@ -0,0 +1,47 @@
+/**
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/ORMGetSessionFactory.java:
```diff
@@ -0,0 +1,47 @@
+/**
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/functions/ORMReload.java:
```diff
@@ -0,0 +1,44 @@
+/**
```
see comment for EntityDelete

michaeloffner commented on source/java/src/org/lucee/extension/orm/hibernate/util/CommonUtil.java:
```diff
@@ -0,0 +1,1068 @@
+package org.lucee.extension.orm.hibernate.util;
```
hard to see the change, when the function not got moved but deleted and created. did you just chnage the package name or also code inside the class. best do not move stuff around at all,this can be done is a separate "clean up PR". 
In short move the class back to it's old package

michaeloffner commented on source/java/src/org/lucee/extension/orm/hibernate/util/ConfigurationBuilder.java:
```diff
@@ -0,0 +1,299 @@
+package org.lucee.extension.orm.hibernate.util;
```
same as CommonUtil i guess

michaeloffner commented on source/java/src/org/lucee/extension/orm/hibernate/util/aprint.java:
```diff
@@ -1,432 +0,0 @@
-/**
```
revert that change, add it again

michaeloffner commented on source/tld/ormTags.tldx:
```diff
@@ -0,0 +1,21 @@
+
```
this tld has no use, why did you add it?

michaeloffner commented on src/test/java/HibernateORMEngineTest.java:
```diff
@@ -0,0 +1,28 @@
+package org.lucee.extension.orm.hibernate;
```
we make all our test cases on CFML level, for one simple reason. we wanna have the test as close to real world as possible, so no java based test cases, please remove

michaeloffner commented:
I have reviewed part of the PR, not everything because we have some issue repeating itself over and over again.
We have 125 files changed but in my opinion the changes that matter are only in 2 or 3 files. Other changes are just noise hard to filter out.

"Noise problems"
you have moved all the orm functions from the Lucee core to the extension, this adds a lot of functions , but it is not clear if you have simply copy/paste them or also made some changes, we need that info. In addition all this functions are useless because they are not registered. In addition they do not implement the BIF class and because of that create a lot of overhead. Also i'm not sure if they load the session in the correct way. But biggest question is why did you do it? I think i know why, we did similar with the S3 extension.

You have changed couple of package names of classes and not moved them. so they appear as new classes and the old one got deleted, so like with the functions it is unclear if you simply changed the package name or also modified the classes, best revert that change and do it in a separate clean up PR.

Please do not remove code like print/aprint, that is used in every Lucee project and do not indroduce new stuff like junit that is not used in any Lucee project, let's just stick to the necessary changes with that PR.

michaelborn commented:
The `ORM*()` and `entity*()` methods were moved into the extension to resolve several main issues:
* The ORM methods can be updated and improved without requiring a Lucee core update.
* The ORM functionality is only utilized within the Hibernate extension, so it makes sense to encapsulate those methods therein.
* Experimental ORM methods can be added to (and removed from!) the Hibernate extension without requiring "buy-in" from the Lucee core.

Besides these concerns, having ORM functionality encapsulated within the Lucee ORM extension makes practical and organizational sense. 

With that said, the `ormFunctions.fld` file was supposed to point to the ported ORM functions - that is my fault! I will update this.

Other feedback items:

* The `source/java/.project` changes were made automatically by VS Code. That file only pertains to the Hibernate extension project - I am not sure why we should care if it is incompatible with another Lucee project. The IDE shouldn't care, as it should simply pick up the file and move forward with the project configuration. _That is the entire reason the `.project` file exists._
* Having all ORM and Entity functions extend BIF will require us to implement the `invoke()` method for each class for switching between `call()` methods based on the number and type of arguments. I am not opposed to this, but it will take some time to update the function definitions.
* Since the `print.java` and `aprint.java` classes are unused in production and provide no production value, I would prefer we not ship them with the Hibernate extension. If dump debugging is to be used in 2023, I would recommend using the `aprint.java` class in the Lucee core.
* The `EXTENSION.MF` is used in the Maven build, which uses the echo-maven-plugin to append manifest entries into a Lucee-compatible extension manifest. Since [Lucee's extension manifest](https://docs.lucee.org/guides/lucee-5/extensions.html) does not adhere to the [Jar File Specification](https://docs.oracle.com/javase/8/docs/technotes/guides/jar/jar.html#JAR_Manifest), this is the only way to build a manifest in Maven that will parse successfully in Lucee.
* I strongly disagree with removing junit tests. YES, we need CFML functionality and syntax to be tested in CFML, but it is a huge benefit to have java-based unit tests which are going to run much faster and have much better integration with the IDE. In addition, it allows us to catch bugs in situations which are very hard or verbose to replicate from CFML.
* Finally, moving the Util classes to a .util package makes structural sense. There were several changes to those util classes apart from the package move, though, mostly related to removing reflection when possible and/or reworking the configuration process as part of the ORM reload performance improvements.

Our only aim here is to improve the quality and maintenance overhead of the Hibernate extension. As long as the changes improve maintainability and do not adversely affect the CFML developer, I don't see why these changes should be backed out of the PR.

michaelborn commented:
> With that said, the `ormFunctions.fld` file was supposed to point to the ported ORM functions - that is my fault! I will update this.

Actually, this file is already committed and _does_ point to the moved functions:

https://github.com/lucee/extension-hibernate/blob/934dddc103189e6e09a0e0a72461d91db05809ec/source/fld/ormFunctions.fld#L15-L29

michaelborn commented:
@michaeloffner I updated the built-ins (`ORM*()` and `entity*()`) to `extends BIF` as you suggested.
I also dropped the `.tld` XML config file for configuring built-in tags, since, as you mentioned, it is unused.

Can we get another review on this?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/lucee/extension-hibernate?pullRequest=36) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/lucee/extension-hibernate?pullRequest=36) before we can accept your contribution.<br/>**2** out of **3** committers have signed the CLA.<br/><br/>:white_check_mark: zspitzer<br/>:white_check_mark: michaelborn<br/>:x: Michael Born<br/><hr/>**Michael Born** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/lucee/extension-hibernate?pullRequest=36) it.</sub>
